### PR TITLE
[Not finished]: Parallelize Distance computation

### PR DESCRIPTION
**Pull request by pfistfl**
In this PR i want to provide the opportunity to parallelize distance computation. 
This is done using [parallelMap](https://github.com/berndbischl/parallelMap)
The current idea is to subset the prediction data into batches and parallelize the computation for different batches. 
Additionally i wanted to add the distance computation methods from [rucrdtw](https://github.com/pboesu/rucrdtw) for dynamic time-warping and euclidean distance computation.

## FIXME's: 
- [x] Add rucrdtw dynamic time warping distance
- [x] Add rucrdtw euclidean distance
- [x] Unittest rucrdtw
- [x] Unittest parallelization
- [x] Parallelization for classiKernel
- [x] Benchmark parallelized vs. non-parallelized version

- [x] Open Question: How does ```rucrdtw::ucrdtw()``` compare to ```dtw::dtw()```

## Benchmark
On an Amazon **c4.2xlarge** instance with 8 vcpus, I get the following results:

### Benchmark parallelization
Parallelization hopefully speeds up computation and might reduce memory usage by allowing to cycle through batches.
```r
library("dtw")
library("classiFunc")
library("parallelMap")
library("microbenchmark")

ncpus = parallel::detectCores()
parallelStartMulticore(ncpus)
x = matrix(rnorm(100*200), nrow = 100)
microbenchmark(
  "parallel" = parallelComputeDistMat(x, method = "DTW", window.type = "sakoechiba", 
    window.size = 30, ncpus = ncpus),
  "singlecore" = computeDistMat(x, method = "DTW", window.type = "sakoechiba",
     window.size = 30),
  times = 10)
```
- Commit: 8202dd5
Unit: seconds

 |      expr   |    min   |     lq    |  mean  |  median    |    uq    |   max | neval | cld|
| --- |--- |--- |--- |--- |--- |--- |--- |--- |
 |  parallel |   6.734901  |6.749286 |6.781069  |6.776937  |6.798617  |6.848054 |   10  |a |
| singlecore |15.309943 |15.689094 |15.830482| 15.833187| 15.980210| 16.384951   | 10   |b|

### Benchmark rucrdtw

```r
x = matrix(rnorm(50*100), nrow = 100)
z1 = microbenchmark(
  "rucrdtw" = computeDistMat(x, method = "rucrdtw", dtwwindow = 0.05),
  "dtw:dtw" = computeDistMat(x, method = "DTW", window.type = "sakoechiba", window.size = 5),
  times = 10
)
```
Unit: milliseconds

|expr    |       min|        lq|      mean|    median|        uq|       max| neval|cld |
  |:-------|---------:|---------:|---------:|---------:|---------:|---------:|-----:|:---|
  |rucrdtw |  355.1945|  357.9028|  359.7006|  359.5331|  361.4983|  365.4603|    10|a   |
  |dtw:dtw | 4118.6532| 4159.5649| 4294.4180| 4265.9840| 4292.2974| 4775.2941|    10|b   |

### CODE REVIEW

**Comment by pfistfl** [comment]
@maierhofert
I am mostly done here! Care to review?

Some unsolved problems are memory consumption and results of rucrdtw package but that can not be solved on this end.

**Comment by pfistfl** [comment]
```r
1. parallelStartSocket / parallelStartMulticore
2. knn mit parallel = TRUE
```

File: R/classiKernel.R [hunk 1]
```diff
@@ -194,7 +194,8 @@ classiKernel = function(classes, fdata, grid = 1:ncol(fdata), h = 1,
 #'   additional arguments to \link{computeDistMat}.
 #' @seealso classiKernel
 #' @export
-predict.classiKernel = function(object, newdata = NULL, predict.type = "response", ...) {
+predict.classiKernel = function(object, newdata = NULL, predict.type = "response",
+  parallel = FALSE, ...) {
```
Comment (pfistfl) [inline R/classiKernel.R]: Document this!

Comment (pfistfl) [reply R/classiKernel.R]: Add example

**Comment by maierhofert** [comment]
@pfistfl I just reviewed your code. Thank you for your amazing work. A couple minor comments:

- [ ] add documentation for new functions/parameters
- [ ] add parameters for number of batches/batch size in predict functions
- [ ] put your name in the DESCRIPTION with role = "aut"

File: R/classiKernel.R (hunk 1 above)
Comment (pfistfl) [reply R/classiKernel.R]: Done

**Comment by pfistfl** [comment]
I have added docs for batches param in ```parallelComputeDistMat()```
Added myself to DESCRIPTION
added DOCS

**Comment by pfistfl** [comment]
Using .Call to interface other package's Rcpp Code is discouraged.
[See here](https://stackoverflow.com/questions/24150185/foreign-function-calls-to-a-different-package-note)
I reverted back to the old style.

**Comment by pfistfl** [comment]
R CMD check results
0 errors | 0 warnings | 0 notes

R CMD check succeeded

File: R/classiKernel.R (hunk 1 above)
Comment (maierhofert) [reply R/classiKernel.R]: Looks good.

**Comment by pfistfl** [comment]
Update: The memory issues in rucrdtw seem to be fixed.

**Pull request merged**